tools/python/google_maps_bookmarks.py

```diff
@@ -2,97 +2,133 @@
         print("4. Unzip the export and look for csv files in the folder Takeout/Saved/")
         print()
-        while True:
-            self.csv_file = input("Insert path to csv file: ")
+        self.input_file = input_file
+        if not path.isfile(self.input_file):
+            raise FileNotFoundError(f"Couldn't find {self.input_file}")
+        if not access(self.input_file, R_OK):
+            raise PermissionError(f"Couldn't read {self.input_file}")
 
-            if not self.csv_file:
-                print("Please provide a csv file" + os.linesep)
-                continue
-            elif not os.path.isfile(self.csv_file):
-                print(f"Couldn't find {self.csv_file}" + os.linesep)
-                continue
-            elif not os.access(self.csv_file, R_OK):
-                print(f"Couldn't read {self.csv_file}" + os.linesep)
+        while True:
+            bookmark_list_name = input("Bookmark list name: ")
+            if not bookmark_list_name:
+                print("Please provide a name" + linesep)
                 continue
             else:
+                self.output_file = bookmark_list_name + "." + output_format
                 break
+            
+        self.places = []
+        self.output_format = output_format
 
+    def convert_timestamp(self, timestamp):
+        date = dateutil_parser.isoparse(timestamp)
+        return date.strftime('%Y-%m-%d %H:%M:%S')
+
+    def get_api_key(self):
         while True:
             self.api_key = input("API key: ")
             if not self.api_key:
-                print("Please provide an API key" + os.linesep)
+                print("Please provide an API key" + linesep)
                 continue
             else:
                 break
+        
+    def get_json(self, url):
+        max_attempts = 3
+        for retry in range(max_attempts):
+            try:
+                response = urllib.request.urlopen(url)
+                return json.load(response)
+            except urllib.error.URLError:
+                print(f"Couldn't connect to Google Maps. Retrying... ({retry + 1}/{max_attempts})")
+                if retry < max_attempts - 1:
+                    continue
+                else:
+                    raise
+                
+    def process_geojson_features(self, geojson):
+        for feature in geojson['features']:
+            geometry = feature['geometry']
+            coordinates = geometry['coordinates']
 
-        while True:
-            bookmark_list_name = input("Bookmark list name: ")
-            if not bookmark_list_name:
-                print("Please provide a name" + os.linesep)
-                continue
-            else:
-                self.kml_file = bookmark_list_name + ".kml"
-                break
-        print()
-        self.places = []
+            properties = feature['properties']
+            location = properties.get('location', {})
```

Comment:
Sometimes the coordinates are set to zero in GeoJSON. See https://github.com/organicmaps/organicmaps/pull/8721#pullrequestreview-2247250563 and the comment below. How do you handle this case?

Reply:
Thanks. As you correctly pointed out, for these cases we can request coordinates from the Google API. I've adapted the code to handle this for when we have a google maps URL with query param `q` (and no coords) or `cid`. I have tested this with my own GeoJSON and it works for all cases.

Reply:
Correction: `cid` does not work in the way I had it, and as of yet I haven't figured out a way to convert those. So I've modified the script to skip those and print a note that the script doesn't yet support `cid`. Not ideal but maybe we can handle that in a follow-on PR.

Reply:
The example here https://github.com/organicmaps/organicmaps/pull/8721#issuecomment-2298220602 has a `ftid`. Does your data has a `cid`?

Anyway, the csv code already converts a `ftid` to coordinates by calling `https://maps.googleapis.com/maps/api/place/details/json?ftid=...`. Maybe the same API works with the `cid` parameter?